cvmfs/revision.py

```diff
@@ -0,0 +1,222 @@
+            return self._get_next_dirent()
+
+    def _fetch_and_push_catalog(self, catalog_mountpoint):
+        current_catalog = self._get_current_catalog().catalog
+        nested_ref      = current_catalog.find_nested_for_path(catalog_mountpoint)
+        if not nested_ref:
+            raise NestedCatalogNotFound(self.revision.repository)
+        new_catalog     = nested_ref.retrieve_from(self.revision.repository)
+        self._push_catalog(new_catalog)
+
+    def _has_more(self):
+        return len(self.catalog_stack) > 0
+
+    def _push_catalog(self, catalog):
+        catalog_iterator = self._CatalogIterator(catalog)
+        self.catalog_stack.append(catalog_iterator)
+
+    def _get_current_catalog(self):
+        return self.catalog_stack[-1]
+
+    def _pop_catalog(self):
+        return self.catalog_stack.pop()
+
+
+class CatalogTreeIterator(object):
+    class _CatalogWrapper:
+        def __init__(self, revision):
+            self.revision        = revision
+            self.catalog           = None
+            self.catalog_reference = None
+
+        def get_catalog(self):
+            if self.catalog is None:
+                self.catalog = self.catalog_reference.retrieve_from(self.revision.repository)
+            return self.catalog
+
+    def __init__(self, revision):
+        root_catalog = revision.retrieve_root_catalog()
+        self.revision    = revision
+        self.catalog_stack = collections.deque()
+        wrapper            = self._CatalogWrapper(self.revision)
+        wrapper.catalog    = root_catalog
+        self._push_catalog_wrapper(wrapper)
+
+    def __iter__(self):
+        return self
+
+    def next(self):
+        if not self._has_more():
+            raise StopIteration()
+        catalog = self._pop_catalog()
+        self._push_nested_catalogs(catalog)
+        return catalog
+
+    def _has_more(self):
+        return len(self.catalog_stack) > 0
+
+    def _push_nested_catalogs(self, catalog):
+        for nested_reference in catalog.list_nested():
+            wrapper = self._CatalogWrapper(self.revision)
+            wrapper.catalog_reference = nested_reference
+            self._push_catalog_wrapper(wrapper)
+
+    def _push_catalog_wrapper(self, catalog):
+        self.catalog_stack.append(catalog)
+
+    def _pop_catalog(self):
+        wrapper = self.catalog_stack.pop()
+        return wrapper.get_catalog()
+
+
+class Revision:
+    """ Wrapper around a CVMFS Repository revision"""
+
+    def __init__(self, repository, tag):
+        self.repository = repository
+        self._tag = tag
+
+    def __str__(self):
+        return '<Revision ' + self.revision_number + ' - ' + self.hash + '>'
+
+    @property
+    def revision_number(self):
+        return self._tag.revision
+
+    @property
+    def name(self):
+        return self._tag.name
+
+    @property
+    def timestamp(self):
+        return self._tag.timestamp
+
+    @property
+    def hash(self):
```

Comment:
By the way: I like the `@property` decorator, but which python version is required for that? I feel that we should at least target to support CentOS 6. Could you please double-check...